### PR TITLE
[feature-update] Created a response class for all modules

### PR DESCRIPTION
**Pull request by ayushanand18**
## Overview
This is a long PR (very long, I believe) and I am really sorry for that. 

**This PR aims to add response class `OBISQueryResult` to every module.**

## Changes introduced
A response class in each module has
+ an `__init__` function (to initialize but no specialized function to do
+ all existing functions as methods to the class
+ `get_search_url` returns OBIS API URL for the query
+ some functions have `get_mapper_url` methods to get OBIS Mapper URL, we identify functions to show a mapper url using a `self.mapper` flag (if true for a method then it has a corresponding Mapper URL).
+ docstrings have also been updated

## Additional Changes
+ Tests have also been updated so this PR has become huge. But each of them are passing and new tests for new methods and an extensive testing to check whether returned url do really exist have been introduced. I believe this will make this package more robust.
+ Docs for `occurrences` module have also been changed just to make it easy for figuring out how things would look after this update. But not for other modules. **I am working on another PR after this to work on updating all the docs. I stopped from doing all that in this PR to curtail PR length which is already very huge.**

Thanks!

### CODE REVIEW

**Comment by ayushanand18** [comment]
Once again, I am sorry for this huge PR.

**Review by 7yl4r — CHANGES_REQUESTED** [review]
I don't like the reuse of the name `OBISQueryResult` for all of these classes.

I also see you have duplicate code for `get_search_url` in several of these classes. You can define the methods once in a parent class or you can define the functions in a common module and attach them, either will remove the duplication.

**Comment by ayushanand18** [comment]
> I don't like the reuse of the name `OBISQueryResult` for all of these classes.
> 
> I also see you have duplicate code for `get_search_url` in several of these classes. You can define the methods once in a parent class or you can define the functions in a common module and attach them, either will remove the duplication.

Thank you for the review! 
+ I will rename the classes in each module something more tailored say `NodesQuery` for nodes module and similarly. 
+ I will add a superclass `OBISQueryResult` in obisutils that is inherited by classes in each module.

**Review by 7yl4r — APPROVED** [review]

**Comment by 7yl4r** [comment]
relevant discussions regarding this reorganization are on #69 and #70

File: docs/occurrences.rst [hunk 1]
```diff
@@ -3,12 +3,32 @@
...
+.. autoclass:: OBISQueryResult
+
+An OBISQueryResult object for fetching occurrence records.
+
+Usage
+#####
+
+.. code-block:: python
+
+    from pyobis.occurrences import OBISQueryResult as OQR
```
Comment (ocefpaf) [inline docs/occurrences.rst:17]: While `numpy as np` and others are well established short versions of imports today, with better autocomplete functionality, these short versions kind of obfuscate the code. In our docs I'd rather we use the full name of the class to make it easier to follow the example along.

File: pyobis/checklist/checklist.py [hunk 2]
```diff
@@ -59,53 +70,55 @@ def list(
...
+            positive numbers.
+        :param geometry: [String] Geometry, formatted as WKT or GeoHash.
+        :param flags: [String] Comma separated list of quality flags which need
+            to be set.
+
+        :return: A dictionary
+
+        Usage::
+
+        from pyobis.checklist import ChecklistQuery as CQR
```
Comment (ocefpaf) [inline pyobis/checklist/checklist.py:116]: Same comment about short versions as above.

File: pyobis/checklist/checklist.py [hunk 3]
```diff
@@ -115,54 +128,56 @@ def redlist(
...
-       from pyobis import checklist as ch
-       ch.newest(scientificname='Abra Alba')
-    """
-    url = obis_baseurl + "checklist/newest"
-    scientificname = handle_arrstr(scientificname)
-    out = obis_GET(
-        url,
-        {
+        }
+        out = obis_GET(
```
Comment (ocefpaf) [inline pyobis/checklist/checklist.py:132]: While the API uses capital letter GET, that is odd in python code. Usually capital letter are used only in classes, not in functions. It is not a big deal and you don't have to change it if this is already established in the code.

File: pyobis/checklist/checklist.py [hunk 4]
```diff
@@ -115,54 +128,56 @@ def redlist(
...
+
+        :return: A dictionary
+
+        Usage::
+
+        from pyobis.checklist import ChecklistQuery as CQR
+        ch = CQR()
+        ch.newest(scientificname='Abra Alba')
+        """
+        OBISQueryResult.url = obis_baseurl + "checklist/newest"
```
Comment (ocefpaf) [inline pyobis/checklist/checklist.py:178]: f-strings are much faster than string additions, and easier to read. Again, no need to change, just a minor suggestion.

```suggestion
        OBISQueryResult.url = f"{obis_baseurl}checklist/newest"
```

File: pyobis/obisutils.py [hunk 5]
```diff
@@ -1,6 +1,8 @@
 """
 Utility functions for internal use across various modules.
 """
+from urllib.parse import urlencode
+
 import requests
 
 obis_baseurl = "https://api.obis.org/v3/"
```
Comment (ocefpaf) [inline pyobis/obisutils.py:8]: My personal preference is to not have the last `/` there so the subsequent use of this string gets more readable. For example: `f"{obis_baseurl}/something"` vs`f"{obis_baseurl}something"`.

**Review by ocefpaf — APPROVED** [review]
Massive change but there isn't much we can do here. They all go together and we need this to be able to test the whole thing. Let's go for it.

File: pyobis/checklist/checklist.py [hunk 6]
```diff
@@ -59,53 +70,55 @@ def list(
...
-
-       from pyobis import checklist as ch
-       ch.redlist(scientificname='Abra Alba')"""
-    url = obis_baseurl + "checklist/redlist"
-    scientificname = handle_arrstr(scientificname)
-    out = obis_GET(
-        url,
-        {
+        }
+        out = obis_GET(
```
[the remark this answers is not in the document]
Comment (ayushanand18) [reply pyobis/checklist/checklist.py:74]: That's a great idea but we need to dig deeper into which functions do we need to pass to child classes and will it be worth it. For eg. functions like `stopifnot()`, `obis_GET` shouldn't be available to the user directly and so are kept inside the utils file. Although this is my opinion, and I am open for your advice.

File: docs/occurrences.rst (hunk 1 above)
Comment (ayushanand18) [reply docs/occurrences.rst:17]: Thank you so much for the review! Should I update those docs/docstrings in another PR?

Comment (ocefpaf) [reply docs/occurrences.rst:17]: Maybe. Definitely don't change this one. If @7yl4r agrees with my suggestion we can send another PR later.

File: pyobis/checklist/checklist.py (hunk 4 above)
Comment (ayushanand18) [reply pyobis/checklist/checklist.py:178]: That's a really great learning for me. Optimizing time is the thing I am reading more of in theoretical CS and I think your practical suggestions are helping me learn more. Thank you!

File: pyobis/obisutils.py (hunk 5 above)
Comment (ayushanand18) [reply pyobis/obisutils.py:8]: Noted. I'll make sure code readability is one of my major priorities in subsequent works.

File: docs/occurrences.rst (hunk 1 above)
Comment (7yl4r) [reply docs/occurrences.rst:17]: Full package names are my preference, even with packages like `numpy`. But I agree that we save that for later.

**Pull request merged**